File: symphony/graph/ent/schema/property.go
// Copyright (c) 2004-present Facebook All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.

package schema

import (
	"github.com/facebookincubator/ent"
	"github.com/facebookincubator/ent/schema/edge"
	"github.com/facebookincubator/ent/schema/field"
	"github.com/facebookincubator/ent/schema/index"
)

// PropertyType defines the property type schema.
type PropertyType struct {
	schema
}

// Fields returns property type fields.
func (PropertyType) Fields() []ent.Field {
	return []ent.Field{
		field.String("type"),
		field.String("name"),
		field.Int("index").
			Optional(),
		field.String("category").
			Optional(),
		field.Int("int_val").
			StructTag(`json:"intValue" gqlgen:"intValue"`).
			Optional(),
		field.Bool("bool_val").
			StructTag(`json:"booleanValue" gqlgen:"booleanValue"`).
			Optional(),
		field.Float("float_val").
			StructTag(`json:"floatValue" gqlgen:"floatValue"`).
			Optional(),
		field.Float("latitude_val").
			StructTag(`json:"latitudeValue" gqlgen:"latitudeValue"`).
			Optional(),
		field.Float("longitude_val").
			StructTag(`json:"longitudeValue" gqlgen:"longitudeValue"`).
			Optional(),
		field.Text("string_val").
			StructTag(`json:"stringValue" gqlgen:"stringValue"`).
			Optional(),
		field.Float("range_from_val").
			StructTag(`json:"rangeFromValue" gqlgen:"rangeFromValue"`).
			Optional(),
		field.Float("range_to_val").
			StructTag(`json:"rangeToValue" gqlgen:"rangeToValue"`).
			Optional(),
		field.Bool("is_instance_property").
			StructTag(`gqlgen:"isInstanceProperty"`).
			Default(true),
		field.Bool("editable").
			StructTag(`gqlgen:"isEditable"`).
			Default(true),
		field.Bool("mandatory").
			StructTag(`gqlgen:"isMandatory"`).
			Default(false),
		field.Bool("deleted").
			StructTag(`gqlgen:"isDeleted"`).
			Default(false),
	}
}

// Edges returns property type edges.
func (PropertyType) Edges() []ent.Edge {
	return []ent.Edge{
		edge.From("properties", Property.Type).
			Ref("type"),
		edge.From("location_type", LocationType.Type).
			Ref("property_types").
			Unique(),
		edge.From("equipment_port_type", EquipmentPortType.Type).
			Ref("property_types").
			Unique(),
		edge.From("link_equipment_port_type", EquipmentPortType.Type).
			Ref("link_property_types").
			Unique(),
		edge.From("equipment_type", EquipmentType.Type).
			Ref("property_types").
			Unique(),
		edge.From("service_type", ServiceType.Type).
			Ref("property_types").
			Unique(),
		edge.From("work_order_type", WorkOrderType.Type).
			Ref("property_types").
			Unique(),
		edge.From("project_type", ProjectType.Type).
			Ref("properties").
			Unique(),
	}
}

// Indexes returns property type indexes.
func (PropertyType) Indexes() []ent.Index {
	return []ent.Index{
		index.Fields("name").
			Edges("location_type").
			Unique(),
		index.Fields("name").
			Edges("equipment_port_type").
			Unique(),
		index.Fields("name").
			Edges("equipment_type").
			Unique(),
		index.Fields("name").
			Edges("link_equipment_port_type").
			Unique(),
		index.Fields("name").
			Edges("work_order_type").
			Unique(),
	}
}

// Property defines the property schema.
type Property struct {
	schema
}

// Fields returns property fields.
func (Property) Fields() []ent.Field {
	return []ent.Field{
		field.Int("int_val").
			StructTag(`json:"intValue" gqlgen:"intValue"`).
			Optional(),
		field.Bool("bool_val").
			StructTag(`json:"booleanValue" gqlgen:"booleanValue"`).
			Optional(),
		field.Float("float_val").
			StructTag(`json:"floatValue" gqlgen:"floatValue"`).
			Optional(),
		field.Float("latitude_val").
			StructTag(`json:"latitudeValue" gqlgen:"latitudeValue"`).
			Optional(),
		field.Float("longitude_val").
			StructTag(`json:"longitudeValue" gqlgen:"longitudeValue"`).
			Optional(),
		field.Float("range_from_val").
			StructTag(`json:"rangeFromValue" gqlgen:"rangeFromValue"`).
			Optional(),
		field.Float("range_to_val").
			StructTag(`json:"rangeToValue" gqlgen:"rangeToValue"`).
			Optional(),
		field.String("string_val").
			StructTag(`json:"stringValue" gqlgen:"stringValue"`).
			Optional(),
	}
}

// Edges returns property edges.
func (Property) Edges() []ent.Edge {
	return []ent.Edge{
		edge.To("type", PropertyType.Type).
			Unique().
			Required().
			StructTag(`gqlgen:"propertyType"`),
		edge.From("location", Location.Type).
			Unique().
			Ref("properties").
			StructTag(`gqlgen:"locationValue"`),
		edge.From("equipment", Equipment.Type).
			Unique().
			Ref("properties").
			StructTag(`gqlgen:"equipmentValue"`),
		edge.From("service", Service.Type).
			Unique().
			Ref("properties").
			StructTag(`gqlgen:"serviceValue"`),
		edge.From("equipment_port", EquipmentPort.Type).
			Unique().
			Ref("properties"),
		edge.From("link", Link.Type).
			Unique().
			Ref("properties"),
		edge.From("work_order", WorkOrder.Type).
			Unique().
			Ref("properties"),
		edge.From("project", Project.Type).
			Ref("properties").
			Unique(),
		edge.To("equipment_value", Equipment.Type).
			Unique(),
		edge.To("location_value", Location.Type).
			Unique(),
		edge.To("service_value", Service.Type).
			Unique(),
	}
}

File: symphony/app/fbcnms-projects/inventory/app/common/Property.js
/**
 * Copyright 2004-present Facebook. All Rights Reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @flow
 * @format
 */

import type {FragmentReference} from 'relay-runtime';
import type {PropertyFormField_property} from '../components/form/__generated__/PropertyFormField_property.graphql';
import type {PropertyType} from './PropertyType';

import DateTimeFormat from './DateTimeFormat.js';
import {toMutablePropertyType} from './PropertyType';

export type Property = {|
  id?: ?string,
  propertyType: PropertyType,

  // one or more of the following potential value fields will have actual data,
  // depending on the property type selected for this property.
  // e.g. for 'email' the stringValue field will be populated
  booleanValue?: ?boolean,
  stringValue?: ?string,
  intValue?: ?number,
  floatValue?: ?number,
  latitudeValue?: ?number,
  longitudeValue?: ?number,
  rangeFromValue?: ?number,
  rangeToValue?: ?number,
  equipmentValue?: ?{id: string, name: string},
  locationValue?: ?{id: string, name: string},
  serviceValue?: ?{id: string, name: string},
|};

export const sortPropertiesByIndex = (a: Property, b: Property) =>
  (a.propertyType.index ?? 0) - (b.propertyType.index ?? 0);

export const getNonInstancePropertyTypes = (
  properties: Array<Property>,
  propertyTypes: Array<PropertyType>,
): Array<PropertyType> => {
  properties = properties || [];
  const propIds = properties.map(x => x.propertyType.id);
  return propertyTypes.filter(type => !propIds.includes(type.id));
};

export const getPropertyValue = (property: Property | PropertyType) => {
  {
    const type = property.propertyType
      ? property.propertyType.type
      : property.type;
    switch (type) {
      case 'date':
      case 'email':
      case 'enum':
      case 'string':
        return property.stringValue;
      case 'datetime_local':
        return DateTimeFormat.dateTime(property.stringValue);
      case 'bool':
        return property.booleanValue != undefined
          ? property.booleanValue.toString()
          : '';
      case 'int':
        return property.intValue;
      case 'float':
        return property.floatValue;
      case 'range':
        return property.rangeFromValue !== null &&
          property.rangeToValue !== null
          ? (property.rangeFromValue ?? '') +
              ' - ' +
              (property.rangeToValue ?? '')
          : '';
      case 'gps_location':
        return property.latitudeValue !== null &&
          property.longitudeValue !== null
          ? (property.latitudeValue ?? '') +
              ', ' +
              (property.longitudeValue ?? '')
          : '';
      /**
       * Since this function accepts either property or property type,
       * we need to check which one we recieved.
       * In the case of PropertyType, there isn't an equipment/location value.
       */
      case 'equipment':
        return property.propertyType ? property.equipmentValue?.name : null;
      case 'location':
        return property.propertyType ? property.locationValue?.name : null;
      case 'service':
        return property.propertyType ? property.serviceValue?.name : null;
    }
  }
};

export const toPropertyInput = (properties: Array<Property>): Array<any> => {
  return properties
    .map(property => ({
      ...property,
      propertyTypeID: property.propertyType.id,
    }))
    .map(propInput => {
      const {propertyType: _, ...newPropInput} = propInput;
      return newPropInput;
    })
    .map(property => {
      if ((property.id && property.id.includes('@tmp')) || property.id == '0') {
        const {id: _, ...newProp} = property;
        return newProp;
      }
      return property;
    })
    .map(property => ({
      ...property,
      equipmentValue: undefined,
      equipmentIDValue: property.equipmentValue?.id ?? null,
      locationValue: undefined,
      locationIDValue: property.locationValue?.id ?? null,
      serviceValue: undefined,
      serviceIDValue: property.serviceValue?.id ?? null,
    }));
};

export const toMutableProperty = (
  immutableProperty: $ReadOnly<
    $Diff<PropertyFormField_property, {$refType: FragmentReference, ...}>,
  >,
): Property => ({
  id: immutableProperty.id,
  propertyType: toMutablePropertyType(immutableProperty.propertyType),
  booleanValue: immutableProperty.booleanValue,
  stringValue: immutableProperty.stringValue,
  intValue: immutableProperty.intValue,
  floatValue: immutableProperty.floatValue,
  latitudeValue: immutableProperty.latitudeValue,
  longitudeValue: immutableProperty.longitudeValue,
  rangeFromValue: immutableProperty.rangeFromValue,
  rangeToValue: immutableProperty.rangeToValue,
  equipmentValue:
    immutableProperty.equipmentValue != null
      ? {
          id: immutableProperty.equipmentValue.id,
          name: immutableProperty.equipmentValue.name,
        }
      : null,
  locationValue:
    immutableProperty.locationValue != null
      ? {
          id: immutableProperty.locationValue.id,
          name: immutableProperty.locationValue.name,
        }
      : null,
  serviceValue:
    immutableProperty.serviceValue != null
      ? {
          id: immutableProperty.serviceValue.id,
          name: immutableProperty.serviceValue.name,
        }
      : null,
});
